### PR TITLE
added new function for computing and plotting angle and torsion distributions

### PR DESCRIPTION
**Pull request by cwalker7**
Here is my proposed replacement for the angle and torsion calculator in foldamers/examples/evaluating_parameter_distributions. The new functions in foldamers/parameters takes in a cgmodel and path to trajectory file, assigns angle and torsion types, computes the values using mdtraj, histograms the values, and saves pdf plots.

- I'd be interested to hear any suggestions for improving the part where the array of all angle particles is split into separate arrays for each type to use with mdtraj.

- We should think about how to generalize the angle typing when we have more than just the 2 types of particles (backbone, sidechain)

- I'm thinking we can also make functions in cgmodel for get_angle_type and get_torsion_type to make this easier. We're already doing this within some of the other functions there.

### CODE REVIEW

File: foldamers/parameters/angle_distributions.py [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+            # This returns an [nframes x n_angles] array
+            ang_val_array = md.compute_angles(traj,ang_sub_arrays[type_str[type]])
+            
+            # Reshape arrays and convert to degrees:  
+            ang_val_array = (180/np.pi)*np.reshape(ang_val_array, (nframes*counts[type],1))
+            
+            # Histogram and plot results:
+            n_out, bin_edges_out = np.histogram(
+                ang_val_array, bins=angle_bin_edges,density=True)
+            pyplot.plot(angle_bin_centers,n_out,'o-r',linewidth=1,markersize=6)
```
Comment (mrshirts) [inline foldamers/parameters/angle_distributions.py:138]: I'd make the plots a separate function.

File: foldamers/parameters/angle_distributions.py [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+            n2 += 1
+        if ang_types[i] == 3:
+            ang3_array[n3,:] = ang_array[i,:]
+            n3 += 1
+        if ang_types[i] == 4:
+            ang4_array[n4,:] = ang_array[i,:]    
+            n4 += 1
+    
+    # Save sub_arrays for each angle type to dictionary
+    ang_sub_arrays = {}
```
Comment (mrshirts) [inline foldamers/parameters/angle_distributions.py:104]: Think about ways to loop over an list.  For example, a list of strings ["bb_bb_bb","bb_bb_sc","bb_sc_sc"], and arrays n and ang_array, syo 

for array in ang_sub_arrays:
            if n[array] > 0:
                    ang_sub_array[array] = ang_array.

File: foldamers/parameters/angle_distributions.py [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+    
+    for i in range(len(angle_list)):
+        ang_array[i,0] = angle_list[i][0]
+        ang_array[i,1] = angle_list[i][1]
+        ang_array[i,2] = angle_list[i][2]
+        at1_type = CGModel.get_particle_type(cgmodel,angle_list[i][0])
+        at2_type = CGModel.get_particle_type(cgmodel,angle_list[i][1])
+        at3_type = CGModel.get_particle_type(cgmodel,angle_list[i][2])
+        
+        if at1_type == 'backbone': # bb-X-X
```
Comment (mrshirts) [inline foldamers/parameters/angle_distributions.py:55]: I would take a look at things like get_torsion_force_constant for the new types.  You probably want to assemble a combinatorial dictionary of all the possible types (taking into account the fact that a-a-a-b and b-a-a-a are the same type, but with reversed angle).  The iterate the operations over this list.

File: foldamers/parameters/angle_distributions.py [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+            # Compute all torsion values in trajectory
+            # This returns an [nframes x n_torsions] array
+            torsion_val_array = md.compute_dihedrals(
+                traj,torsion_sub_arrays[type_str[type]])
+            
+            # Reshape arrays and convert to degrees:  
+            torsion_val_array = (180/np.pi)*np.reshape(torsion_val_array, (nframes*counts[type],1))
+            
+            # Histogram and plot results:
+            n_out, bin_edges_out = np.histogram(
```
Comment (mrshirts) [inline foldamers/parameters/angle_distributions.py:296]: Consider making a single graph with subplots for each type?

**Comment by mrshirts** [comment]
Added some comments for now.  Would you be all right with holding off until the general heteropolymer code is updated, and write it to adapt to that?  I'm happy to brainstorm ideas for doing that.

**Comment by cwalker7** [comment]
Ok sure, I'll hold off on this for now - I agree it makes sense to do the general heteropolymer update first.

**Comment by mrshirts** [comment]
Let me know when this is ready to look over again now that #37 is in.

**Comment by cwalker7** [comment]
**A bunch of updates to work with the general heteropolymer cgmodel:**

- Angle and torsion type strings are added as dictionary entries when they are found as a new type. This gets rid of all the logical statements going atom by atom in each angle or torsion, and it works for any amount of particle types. For now I left a debugging print statement telling when a new type is found.

- Plots are moved to a separate function (replacing the old plot_distribution in foldamers.utilities.plot), and they are organized into subplots.

- Switched to a more complicated example which has the structure of get_heteropolymer_1_structure.py in foldamers generating_random_structures examples (3 monomer types each with their own particle types)

**Questions:**

- Do we want to be able to plot angles and torsions by using a more general backbone/sidechain scheme, instead of viewing all particle types as different (for example, a_bb-b_bb-c_bb gets its own type, vs. the same given the broader type bb-bb-bb) 

**TODO:**

- Fix the formatting of the suplots - it is messy when we have more than 2 types

**Comment by mrshirts** [comment]
This looks reasonable. 

> Do we want to be able to plot angles and torsions by using a more general backbone/sidechain scheme, instead of viewing all particle types as different (for example, a_bb-b_bb-c_bb gets its own type, vs. the same given the broader type bb-bb-bb)

I think we can probably handle this with naming backbone/particle bonds differently.  It doesn't seem like a short-time issue.

> Fix the formatting of the suplots - it is messy when we have more than 2 types

Maybe there should be some sort of filter that one only prints some of the subplots (if there are a lot), or prints them out on separate pages as separate plots, so that it doesn't get too crowded if there happen to be more types.

**Comment by cwalker7** [comment]
Ok I will make a separate issue for the subplot formatting, and go ahead and merge this.

**Pull request merged**